### ISSUE TITLE
Failure parsing trait static methods

### ISSUE THREAD

**Issue by danog** [vimeo/psalm#2938]
Psalm does not seem to detect trait static methods: https://psalm.dev/r/10d95a3c4d

**Comment by weirdan** [comment]
I am surprised, but PHP actually allows them to be called without a class: https://3v4l.org/YPQNO

**Comment by muglug** [comment]
Wtf PHP

**Comment by muglug** [comment]
This doesn't feel right to me – @TysonAndre, @OndrejMirtes does this feel like something PHP should allow?

**Comment by muglug** [comment]
For example there's a fatal error here that's near-impossible to discover:

```php
<?php declare(strict_types = 1);

trait T {
    public static function callAbstract() : void {
        self::bar();
    }
    
    abstract static function bar() : void;
}

class A {
    use T;
    
    public static function bar() : void {
        echo "A::bar() called succesfully\n";
    }
}

A::callAbstract(); // fine
T::callAbstract(); // fatal error
```

**Comment by ondrejmirtes** [comment]
Yeah, I'm also gonna implement the check to disallow calling static methods on traits. PHPStan doesn't analyse traits in isolation (only as part of classes that use them) so it cannot allow calling them in isolation either, to prevent the problem with `callAbstract`.

Thanks for pointing it out @danog although the outcome is probably something you didn't expect :D

**Comment by danog** [comment]
<3

**Comment by TysonAndre** [comment]
Psalm also fails to detect when calling an abstract method directly - https://psalm.dev/r/7bfa0b11f7

This is a separate issue, though - In a trait, `self` would normally refer to the class using the trait, not the trait itself.

Also, Psalm (and Phan) fails to warn about calling an abstract static method, in general. I assume that's because there would still be false positives with `$subtypeOfBase::abstractMethod()` or `static::abstractMethod()`

```php
<?php declare(strict_types = 1);

abstract class Base {
    public static function callAbstract() : void {
        self::bar();
    }

    abstract static function bar() : void;
}

class A extends Base {
    public static function bar() : void {
        echo "A::bar() called succesfully\n";
    }
}

A::callAbstract(); // fine
Base::callAbstract(); // fatal error
Base::bar();  // fatal error, currently no warning
```

**Comment by ondrejmirtes** [comment]
Just tested PHPStan, at least it detects those :)

**Comment by TysonAndre** [comment]
https://psalm.dev/r/7bfa0b11f7 is lower priority and should also be a warning, but apparently php allows calling abstract __callStatic methods.

**Comment by muglug** [comment]
> Psalm also fails to detect when calling an abstract method directly

That's fixed in 0d62fbd

**Comment by muglug** [comment]
closing as a wontfix
